### PR TITLE
Foldfix

### PR DESCRIPTION
**Pull request by bmanga**
Reworked some of the code. Now some simple transformations are correctly applied.
Although I have tried some basic examples and they work correctly, this should require some more careful testing to make sure nothing broke.

### CODE REVIEW

**Comment by bmanga** [comment]
This closes issue #340 .

**Comment by bmanga** [comment]
I couldn't find any regressions for this patch.

**Comment by jmalak** [comment]
Give me enough time, because it is very hard to follow your changes.
It is as new code, that I will need to study it in more details.
Anyway, better is to modify existing code by minimal way until it is necessary to do massive code changes.

**Comment by bmanga** [comment]
I rely on the constant folding to actually work for a following pull request implementing nullptr conversions to bool, that's why I had to fix this.

**Comment by jmalak** [comment]
Submitted as squashed commit bb4a67aac3a78f7e5392b85e4944d8b8fb156219
Thanks for your contribution
